doc/developer/design/20210831_correctness.md

```diff
@@ -0,0 +1,124 @@
+# <Insert Name of Feature>
+
+## Summary
+
+A description of the correctness properties Materialize provides and the terms and concepts we use to describe them.
+
+## Goals
+
+The main goal is to lay the foundation for ongoing work to clearly articulate correctness properties of Materialize.
+This includes the introduction of concepts that underlie correctess, the components that will enforce them, and how they can communicate clearly.
+
+## Non-Goals
+
+It is not a goal to prescribe fine detail in how each of the components should behave, or which should have certain responsibilities.
+It is not a goal to prescribe which behavior Materialize should support, but only how to speak about a "correct" subset of it.
+It is not a goal to advocate (much) for DEFINITE over NON-VOLATILE.
+
+## Description
+
+Materialize means to provide several correctness guarantees, which manifest as various consistency and durability properties.
+At the heart of this is the idea that the ordering of events can be explicitly resolved and recorded, after which queries and views over the data become a matter of correct evaluation on completed subsets of time.
+The guarantees of Materialize derive from the durability of these recorded data, and the determinism of views over them.
+
+The primary term introduced here is "definite" to modify a collection of data.
+A collection is definite if it is uniquely named or defined such that anyone with the same identification will see the same collection of data at the same logical times.
+A definite collection has had its updates put in a specific order, by way of the timestamps on updates.
+A handle to a collection can be "definite from `since` until `upper`" (names pending) if it reflects the definite results for times greater or equal to `since` but not greater or equal to `upper`.
+Among the correctness guarantees of Materialize, a main one will be statements of definiteness over intervals of times for collections it maintains.
+
+### Components
+
+There are three components to discuss: storage, compute, and coordination.
+
+The **storage layer** (`persist`, plus potentially others) presents streams of updates to collections to other parts of the system (primary the compute layer).
```

Comment:
Is it correct to say that in this definition, the "sources" (including e.g. decode) lives in the **storage layer** ?

Reply:
That's my mental model, for sure. At least, of { storage, dataflow, coordinator } they seem most closely bound to storage. Perhaps there is a better term that sounds more stream-y, but to the extent that "storage" is the name for what owns access to data, and properties of that data, then yes.